src/evaluators/JSXElement.js

```diff
@@ -215,104 +194,113 @@ function evaluateJSXChildren(
+  let abstractPropsArgs = [];
+  let containsAbstractSpreadAttribute = false;
+  let spreadAttributesCount = 0;
+  let spreadAttributesWithInitialPropsHintCount = 0;
+  let attributesAssigned = 0;
+  let spreadValue;
+
+  const setConfigProperty = (name: string, value: Value): void => {
+    invariant(config instanceof ObjectValue);
+    Properties.Set(realm, config, name, value, true);
+    attributesAssigned++;
+  };
 
-      if (defaultPropValue instanceof Value) {
-        if (key === "children") {
-          if (children === null) {
-            children = defaultPropValue;
-          }
-        } else {
-          attributes.set(key, defaultPropValue);
-        }
-      }
-    }
-  }
   for (let astAttribute of astAttributes) {
     switch (astAttribute.type) {
       case "JSXAttribute":
         let { name, value } = astAttribute;
 
         invariant(name.type === "JSXIdentifier", `JSX attribute name type not supported: ${astAttribute.type}`);
-        attributes.set(name.name, evaluateJSXValue(((value: any): BabelNodeJSXIdentifier), strictCode, env, realm));
+        setConfigProperty(name.name, evaluateJSXValue(((value: any): BabelNodeJSXIdentifier), strictCode, env, realm));
         break;
       case "JSXSpreadAttribute":
-        let spreadValue = Environment.GetValue(realm, env.evaluate(astAttribute.argument, strictCode));
+        spreadValue = Environment.GetValue(realm, env.evaluate(astAttribute.argument, strictCode));
 
-        if (spreadValue instanceof ObjectValue) {
-          for (let [key, spreadProp] of spreadValue.properties) {
-            if (spreadProp !== undefined && spreadProp.descriptor !== undefined) {
-              let spreadPropValue = spreadProp.descriptor.value;
-
-              if (spreadPropValue instanceof Value) {
-                if (key === "children") {
-                  children = spreadPropValue;
-                } else {
-                  attributes.set(key, spreadPropValue);
-                }
-              }
+        if (spreadValue instanceof ObjectValue && !spreadValue.isPartialObject()) {
+          for (let [spreadPropKey, binding] of spreadValue.properties) {
+            if (binding && binding.descriptor) {
+              setConfigProperty(spreadPropKey, Get(realm, spreadValue, spreadPropKey));
             }
           }
         } else {
-          throw new FatalError("ObjectValues are the only supported value for JSX Spread Attributes");
+          spreadAttributesCount++;
+          containsAbstractSpreadAttribute = true;
+          invariant(spreadValue instanceof AbstractValue || spreadValue instanceof ObjectValue);
+
+          // we push the props up to this point into the abstract props args. we also
+          // push the abstract spread object and then we create a fresh props object
+          if (objectHasNoPartialKeyAndRef(realm, spreadValue)) {
+            spreadAttributesWithInitialPropsHintCount++;
+            if (config.properties.size > 2) {
```

Comment:
It's 2 because an empty object that has a deleted key and ref, is two properties in size. It looks confusing, but makes sense if maybe I added a comment? I'll try and address your other point by moving the code around a bit too.

Reply:
Maybe we can compare property names instead? That would be more explicit.

It would also help to extract config creation into a separate function that immediately deletes key and ref. Then we don't need to "remember" to delete them in two different places (for first and subsequent configs). I think it would also be clearer if `config` was renamed to `partialConfig` or something that better represents the fact that we may create *many* such configs as we go.

In either case, I still don't quite understand why we need to use `2` here but `0` just below.

Reply:
What about the implementation I made in the follow commit? That okay?